### PR TITLE
WIP: MLPerf Training v5.1 Flux.1 text-to-image [bounty]

### PR DESCRIPTION
**Pull request by UbajakaCJ**
# [WIP] MLPerf Training v5.1: Flux.1 text-to-image

Closes / references: **Bounty — mlperf Training flux** (row 5 in the tinygrad bounties sheet)

---

## What this PR adds

| File | Purpose |
|---|---|
| `extra/models/flux.py` | Full Flux.1 DiT architecture in native tinygrad |
| `examples/mlperf/model_train.py` | `train_flux()` training loop |
| `examples/mlperf/dataloader.py` | `batch_load_train_flux()` for pre-encoded CC12M |

---

## Architecture (Flux.1)

MLPerf Training v5.1 replaces Stable Diffusion v2 with **Flux.1**, a 11.9B-parameter
Diffusion Transformer. Key differences from the existing SDv2 implementation:

- **DiT backbone** instead of UNet — double-stream blocks (separate img+txt) → single-stream blocks (fused)
- **Rectified flow loss** (flow-matching) instead of DDPM v-prediction
  - `x_t = (1-t)*noise + t*data`, target `v = data - noise`, `t ~ logit-normal(0,1)`
- **2D RoPE** on image patch tokens (not text tokens)
- **Pre-computed latents + embeddings** — VAE, T5-XXL, CLIP all frozen; all encodings are
  offline pre-computed exactly as in the SDv2 benchmark
- **Validation metric: validation loss** (not FID/CLIP), which makes eval fast and avoids
  generating 30K images during the timed training window

MLPerf config:
```
in_channels=16, hidden_size=3072, n_heads=24
double_stream_layers=19, single_stream_layers=38
txt_in_dim=4096 (T5-XXL), vec_in_dim=768 (CLIP)
patch_size=2
dataset: 1,099,776 CC12M samples
```

---

## Current status (WIP — locking bounty)

- [x] Full model architecture: `FluxModel`, `DoubleStreamBlock`, `SingleStreamBlock`, `LastLayer`
- [x] 2D RoPE for image patches (`build_2d_rope_freqs`, `apply_rope_2d`)
- [x] `TimestepEmbedding`, `VectorEmbedding`, modulated LayerNorm (`AdaLN`)
- [x] Rectified flow loss (`rectified_flow_loss`)
- [x] `train_flux()` training loop with multi-GPU sharding, checkpointing, validation
- [x] `batch_load_train_flux()` dataloader skeleton for pre-encoded CC12M webdataset
- [ ] Verify dataset key names against mlcommons/training flux1/ reference
- [ ] Load pre-trained Flux.1 starting checkpoint (required for MLPerf closed division)
- [ ] Numerical validation of forward pass against HuggingFace diffusers FluxPipeline
- [ ] BF16 autocast tuning + BEAM search for 6×7900XTX (tinybox)
- [ ] Hit val_loss quality target within 24h wall-clock

---

## How to run (once dataset is ready)

```bash
# Download pre-encoded CC12M dataset from mlcommons-storage (TBD once reference is confirmed)
# python3 extra/datasets/cc12m_flux.py --output ./datasets/cc12m

# Train (single GPU test)
MODEL=flux DATADIR=./datasets/cc12m FLUX_CKPTDIR=./checkpoints/flux \
  BS=4 GPUS=1 python3 examples/mlperf/model_train.py

# Train (tinybox 6×7900XTX)
MODEL=flux DATADIR=./datasets/cc12m FLUX_CKPTDIR=./checkpoints/flux \
  BS=24 GPUS=6 DEFAULT_FLOAT=HALF TRAIN_BEAM=4 BEAM_UOPS_MAX=1500 \
  python3 examples/mlperf/model_train.py
```

---

## Notes

The 2D RoPE implementation follows the Black Forest Labs original Flux repo and applies
`D//4`-dim frequency vectors independently for H and W, interleaved across the head dim.
This differs from 1D NLP RoPE (the text tokens do not get RoPE applied).

The rectified flow timestep sampling uses **logit-normal** (sigmoid of standard normal),
matching the Flux.1 training procedure in *Scaling Rectified Flow Transformers for
High-Resolution Image Synthesis* (Esser et al. 2024).

### CODE REVIEW

**Comment by UbajakaCJ** [comment]
image
Submitting this draft to lock: mlperf Training flux -- $1000
